### PR TITLE
🐛 BSLs with validation disabled should be validated at least once

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

### CODE REVIEW

**Comment by carlisia** [comment]
> BSLs with validation disabled should be validated at least once

Why?

**Comment by ashish-amarnath** [comment]
> Why?

Because without that the rule at https://github.com/vmware-tanzu/velero/blob/main/internal/storage/storagelocation.go#L44 is not satisfied and the status of BSLs whose repeated validation is disabled will never be set to available.

**Review by carlisia — CHANGES_REQUESTED** [review]
Glad you caught this discrepancy!

I made some suggestions/requests.

File: internal/storage/storagelocation.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation.go:73]: I found `!time.Now().UTC().Before(nextValidation) ` to be more readable!
Read as "Time now is not before next validation"

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go:45]: I chose to use the variable names as they were more descriptive and the reason for test failure can be easily narrowed down.
The reason for this change is that the test was incorrect. We want this method to return `true` in this test but the previously expected result was `false`. Also, the diff seems misaligned.

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go:53]: words are hard

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go:45]: Also, preferring the "should" naming convention as it is more direct in what should happen, and when the test fails we will know exactly what should have happened.

File: internal/storage/storagelocation.go
Comment (nrb) [reply internal/storage/storagelocation.go:73]: Readability can be a subjective measure, so I think saying one is more readable than the other universally is tough.

I find the expanded version proposed easier to read and more obvious on first pass, myself.

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go]: Fixed.

File: internal/storage/storagelocation.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation.go]: Updated to use the expanded version.

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go]: @carlisia I am marking this as resolved. Please tell me if you feel strongly about rewording this.

File: internal/storage/storagelocation_test.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go:42]: Marking this as resolved.
@carlisia Please unresolve it if you feel strongly that this needs to be changed.

File: internal/storage/storagelocation.go
Comment (nrb) [inline internal/storage/storagelocation.go:65]: Since GitHub doesn't let me comment on unchanged lines, I think we should mention this special case up in the function's comment from L41-L46 as one of the rules. Based on the comment on L44, one might expect `0` to to return `true`.

**Review by nrb — CHANGES_REQUESTED** [review]
Aside from one minor comment change, I think I'm 👍  on this.

File: internal/storage/storagelocation.go
Comment (ashish-amarnath) [reply internal/storage/storagelocation.go:65]: I think L43 and L44 together define the expected behavior. The bug is that the code didn't match the comments.

**Review by nrb — DISMISSED** [review]

**Review by zubron — DISMISSED** [review]
Thanks, @ashish-amarnath!

**Comment by zubron** [comment]
Looks like this needs a final review and approval from @carlisia before merging.

File: internal/storage/storagelocation_test.go
Comment (carlisia) [reply internal/storage/storagelocation_test.go:42]: I really would like to have the changes I mentioned, mainly, not to introduce a new way to name tests (we consistently use an active voice instead of "should"), and avoid using variable names in the test name that we'd have to change if we rename the variable. Thank you! This is the only change I'd like to see for this and then it's ready to merge.

Comment (ashish-amarnath) [reply internal/storage/storagelocation_test.go]: Done!

**Review by zubron — APPROVED** [review]

**Review by carlisia — APPROVED** [review]
Thank you for this 🙏

**Pull request merged**
